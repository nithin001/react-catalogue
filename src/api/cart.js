import instance from './base';
import Ajv from 'ajv';

export const get = () => {
  return new Promise((resolve, reject) => {
    instance.get('/cart')
      .then((response) => {
        if (response.data === undefined) {
          reject('backend_error');
          return;
        }
        const data = response.data;
        var ajv = new Ajv();
        var validate = ajv.compile(get_schema);
        var valid = validate(data);
        if (valid) {
          resolve(data);
        } else {
          reject('backend_error');
        }
      })
      .catch(error => {
        reject('backend_error');
      });
  });
};

export const put = (request) => {
  return new Promise((resolve, reject) => {
    instance.put('/cart', { data: request })
      .then((response) => {
        if (response.data === undefined) {
          reject('backend_error');
          return;
        }
        const data = response.data;
        var ajv = new Ajv();
        var validate = ajv.compile(put_schema);
        var valid = validate(data);
        if (valid) {
          resolve(data);
        } else {
          reject('backend_error');
        }
      })
      .catch(error => {
        reject('backend_error');
      });
  });
};

const get_schema = {
  '$id': 'http://challenge.monoqi.net/cart',
  'type': 'object',
  'definitions': {},
  '$schema': 'http://json-schema.org/draft-07/schema#',
  'properties': {
    'lines': {
      '$id': '/properties/lines',
      'type': 'array',
      'items': {
        '$id': '/properties/lines/items',
        'type': 'object',
        'properties': {
          'sku': {
            '$id': '/properties/lines/items/properties/sku',
            'type': 'string',
            'title': 'The Sku Schema ',
            'default': '',
            'examples': [
              '199203'
            ]
          },
          'quantity': {
            '$id': '/properties/lines/items/properties/quantity',
            'type': 'integer',
            'title': 'The Quantity Schema ',
            'default': 0,
            'examples': [
              3
            ]
          }
        },
        'required': ['sku', 'quantity']
      }
    }
  },
  'required': ['lines']
};

const put_schema = {
  '$id': 'http://example.com/example.json',
  'type': 'object',
  'definitions': {},
  '$schema': 'http://json-schema.org/draft-07/schema#',
  'properties': {
    'lines': {
      '$id': '/properties/lines',
      'type': 'array',
      'items': {
        '$id': '/properties/lines/items',
        'type': 'object',
        'properties': {
          'sku': {
            '$id': '/properties/lines/items/properties/sku',
            'type': 'string',
            'title': 'The Sku Schema ',
            'default': '',
            'examples': [
              '199203'
            ]
          },
          'name': {
            '$id': '/properties/lines/items/properties/name',
            'type': 'string',
            'title': 'The Name Schema ',
            'default': '',
            'examples': [
              'IPad'
            ]
          },
          'price': {
            '$id': '/properties/lines/items/properties/price',
            'type': 'object',
            'properties': {
              'amount': {
                '$id': '/properties/lines/items/properties/price/properties/amount',
                'type': 'integer',
                'title': 'The Amount Schema ',
                'default': 0,
                'examples': [
                  229
                ]
              },
              'currency': {
                '$id': '/properties/lines/items/properties/price/properties/currency',
                'type': 'string',
                'title': 'The Currency Schema ',
                'default': '',
                'examples': [
                  'EUR'
                ]
              }
            },
            'required': [
              'amount',
              'currency'
            ]
          },
          'quantity': {
            '$id': '/properties/lines/items/properties/quantity',
            'type': 'integer',
            'title': 'The Quantity Schema ',
            'default': 0,
            'examples': [
              3
            ]
          },
          'lineTotal': {
            '$id': '/properties/lines/items/properties/lineTotal',
            'type': 'object',
            'properties': {
              'amount': {
                '$id': '/properties/lines/items/properties/lineTotal/properties/amount',
                'type': 'integer',
                'title': 'The Amount Schema ',
                'default': 0,
                'examples': [
                  687
                ]
              },
              'currency': {
                '$id': '/properties/lines/items/properties/lineTotal/properties/currency',
                'type': 'string',
                'title': 'The Currency Schema ',
                'default': '',
                'examples': [
                  'EUR'
                ]
              }
            },
            'required': [
              'amount',
              'currency'
            ]
          }
        },
        'required': [
          'sku',
          'name',
          'price',
          'quantity',
          'lineTotal'
        ]
      }
    },
    'total': {
      '$id': '/properties/total',
      'type': 'object',
      'properties': {
        'amount': {
          '$id': '/properties/total/properties/amount',
          'type': 'integer',
          'title': 'The Amount Schema ',
          'default': 0,
          'examples': [
            1126
          ]
        },
        'currency': {
          '$id': '/properties/total/properties/currency',
          'type': 'string',
          'title': 'The Currency Schema ',
          'default': '',
          'examples': [
            'EUR'
          ]
        }
      },
      'required': [
        'amount',
        'currency'
      ]
    }
  },
  'required': [
    'lines',
    'total'
  ]
};