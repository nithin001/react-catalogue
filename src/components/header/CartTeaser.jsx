import React from 'react';
import { connect } from 'react-redux';
export const CartTeaser = ({ loading, error, count, amount, currency }) => {
  let text = '';
  if (loading) {
    text = <span>Loading</span>;
  } else if (error) {
    text = <span>Error loading cart</span>;
  } else {
    text = <div>
      <span className={'count'}>{count}</span>
      <span className={'amount'}>{amount}</span>
      <span className={'currency'}>{currency}</span>
    </div>;
  }
  return <div className={'cart-teaser'}>{text}</div>;
};

export const mapStateToProps = (state) => {
  const error = state.ui.get('error');
  if (error) {
    return { error };
  }
  const loading = state.ui.get('cart_loading') || state.ui.get('quote_loading');
  if (loading) {
    return { loading };
  }

  const count = state.cart.size;
  const amount = state.quote.getIn(['total', 'amount']);
  const currency = state.quote.getIn(['total', 'currency']);
  return { loading, error, count, amount, currency };
};

export default connect(mapStateToProps, null)(CartTeaser);


