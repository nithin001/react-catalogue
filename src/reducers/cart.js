import Immutable from 'immutable';
const cart = (state = Immutable.Map(), action) => {
  switch (action.type) {
    case 'POPULATE_CART': {
      const items = action.payload.lines;
      return items.reduce((result, item) => {
        return result.set(item.sku, item.quantity);
      }, Immutable.Map());
    }
    case 'ADD_TO_CART': {
      const sku = action.payload;
      if (state.has(sku)) {
        const quantity = state.get(sku);
        return state.set(sku, quantity + 1);
      }
      return state.set(sku, 1);
    }
    case 'REMOVE_FROM_CART': {
      const sku = action.payload;
      const quantity = state.get(sku);
      if (quantity === 1) {
        return state.delete(sku);
      }
      return state.set(sku, quantity - 1);
    }
  }
  return state;
};
export default cart;