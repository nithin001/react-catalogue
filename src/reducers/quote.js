import Immutable from 'immutable';
const quote = (state = Immutable.Map(), action) => {
  if (action.type === 'POPULATE_QUOTE') {
    return Immutable.fromJS(action.payload);
  }
  return state;
};
export default quote;