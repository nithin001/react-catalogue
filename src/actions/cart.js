import { get } from '../api/cart';
export const load = () => {
  return (dispatch) => {
    dispatch({ type: 'LOAD_CART_START' });
    return get()
      .then((cart) => {
        dispatch({ type: 'POPULATE_CART', payload: cart });
        dispatch({ type: 'LOAD_CART_COMPLETE' });
      })
      .catch(() => {
        dispatch({ type: 'LOAD_ERROR' });
      });
  };
};