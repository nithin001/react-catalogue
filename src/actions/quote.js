import { put } from '../api/cart';
export const load = (cart) => {
  return (dispatch) => {
    dispatch({ type: 'LOAD_QUOTE_START' });
    return put(cart)
      .then((quote) => {
        dispatch({ type: 'POPULATE_QUOTE', payload: quote });
        dispatch({ type: 'LOAD_QUOTE_COMPLETE' });
      })
      .catch(() => {
        dispatch({ type: 'LOAD_ERROR' });
      });
  };
};