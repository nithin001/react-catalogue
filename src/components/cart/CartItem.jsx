import React from 'react';
import { connect } from 'react-redux';
import { QuantityBar } from './QuantityBar';
import { addToCart, removeFromCart } from '../../actions/cart';
export const CartItem = ({ sku, price, name, quantity, addToCart, removeFromCart }) => {
  return <div className={'cart-item'}>
    <span className={'name'}>{name}</span>
    <span className={'price-amount'}>{price.amount}</span>
    <span className={'price-currency'}>{price.currency}</span>
    <QuantityBar sku={sku} quantity={quantity} addToCart={addToCart} removeFromCart={removeFromCart}/>
  </div>;
};

export const mapStateToProps = (state, ownProps) => {
  const article = state.catalog.filter(item => {
    return item.get('sku') === ownProps.sku;
  }).first();
  const price = {
    amount: article.getIn(['price', 'amount']),
    currency: article.getIn(['price', 'currency']),
  };
  return { price, name: article.get('name') };
};

export const mapDispatchToProp = (dispatch) => ({
    addToCart: (sku) => dispatch(addToCart(sku)),
    removeFromCart: (sku) => dispatch(removeFromCart(sku)),
  }
);

export default connect(mapStateToProps, mapDispatchToProp)(CartItem);
