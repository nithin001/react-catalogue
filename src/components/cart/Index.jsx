import React from 'react';
import { connect } from 'react-redux';
import CartItem from './CartItem';
export const Cart = ({ cart }) => {
  const skus = [];
  cart.keySeq().forEach(key => skus.push(key));
  const items = skus.map((key) => (<CartItem
    key={key}
    sku={key}
    quantity={cart.get(key)}
  />));
  return <div className={'cart'}>{items}</div>;
};

export const mapStateToProps = (state) => {
  return { cart: state.cart };
};

export default connect(mapStateToProps, null)(Cart);
