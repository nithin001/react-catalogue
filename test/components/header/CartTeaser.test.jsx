import React from 'react';
import renderer from 'react-test-renderer';
import Immutable from 'immutable';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { CartTeaser, mapStateToProps } from '../../../src/components/header/CartTeaser';
import cart from '../../../src/reducers/cart';
import quote from '../../../src/reducers/quote';
configure({ adapter: new Adapter() });

describe('CartTeaser component', () => {
  it('should render the CartTeaser component with loading text', () => {
    const props = { loading: true, error: false };
    const title = renderer.create(<CartTeaser
      {...props}
    />);
    expect(title).toMatchSnapshot();
  });

  it('should render the CartTeaser component with error text', () => {
    const props = { loading: false, error: true };
    const title = renderer.create(<CartTeaser
      {...props}
    />);
    expect(title).toMatchSnapshot();
  });

  it('should render the CartTeaser component with count and amount', () => {
    const props = { loading: false, error: false, count: 5, amount: 60, currency: 'EUR' };
    const title = renderer.create(<CartTeaser
      {...props}
    />);
    expect(title).toMatchSnapshot();
  });
});

describe('Connected CartTeaser component', () => {

  it('should map state values to props of the component', () => {
    var mockCart = {
      lines: [{
        'sku': '199203',
        'quantity': 3
      }]
    };

    var mockQuote = {
      'lines': [{
        'sku': '374847',
        'name': 'Nice Product',
        'price': {
          'amount': 20,
          'currency': 'EUR'
        },
        'quantity': 3,
        'lineTotal': {
          'amount': 60,
          'currency': 'EUR'
        }
      }],
      'total': {
        'amount': 60,
        'currency': 'EUR'
      }
    };
    const cartState = cart(undefined, { type: 'POPULATE_CART', payload: mockCart });
    const quoteState = quote(undefined, { type: 'POPULATE_QUOTE', payload: mockQuote });
    const state = {
      ui: Immutable.Map()
        .set('cart_loading', false)
        .set('quote_loading', false)
        .set('error', false),
      cart: cartState,
      quote: quoteState,
    };
    const result = mapStateToProps(state);
    expect(result).toEqual({ loading: false, error: false, count: 1, amount: 60, currency: 'EUR' });
  });

  it('should return only loading when cart is loading', () => {
    const state = {
      ui: Immutable.Map()
        .set('cart_loading', true)
        .set('quote_loading', false)
        .set('error', false),
    };
    const result = mapStateToProps(state);
    expect(result).toEqual({ loading: true });
  });

  it('should return only loading when quote is loading', () => {
    const state = {
      ui: Immutable.Map()
        .set('cart_loading', false)
        .set('quote_loading', true)
        .set('error', false),
    };
    const result = mapStateToProps(state);
    expect(result).toEqual({ loading: true });
  });

  it('should return only error when there is an error', () => {
    const state = {
      ui: Immutable.Map()
        .set('error', true),
    };
    const result = mapStateToProps(state);
    expect(result).toEqual({ error: true });
  });
});