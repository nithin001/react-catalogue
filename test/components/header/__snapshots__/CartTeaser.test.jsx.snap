// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartTeaser component should render the CartTeaser component with count and amount 1`] = `
<div
  className="cart-teaser"
>
  <div>
    <span
      className="count"
    >
      5
    </span>
    <span
      className="amount"
    >
      60
    </span>
    <span
      className="currency"
    >
      EUR
    </span>
  </div>
</div>
`;

exports[`CartTeaser component should render the CartTeaser component with error text 1`] = `
<div
  className="cart-teaser"
>
  <span>
    Error loading cart
  </span>
</div>
`;

exports[`CartTeaser component should render the CartTeaser component with loading text 1`] = `
<div
  className="cart-teaser"
>
  <span>
    Loading
  </span>
</div>
`;
